core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/JniPCConnector.java

```diff
@@ -126,15 +140,94 @@ private static boolean initNativeCode() {
+     * @return true on success, otherwise false
+     * @throws IOException If there are errors in opening/writing/reading/closing etc.
+     *         Note that the method might throw RuntimeExceptions due to critical issues
+     */
+    private static boolean loadLibrary() throws IOException {
         String model = System.getProperty("sun.arch.data.model");
         String libraryToLoad = NATIVE_LIBRARY_64;
         if (!BITS_MODEL_64.equals(model)) {
             return false;
         }
 
-        System.loadLibrary(libraryToLoad);
+        File dllPath = buildDllLocalPath();
+
+        File dllOnDisk = new File(dllPath + File.separator + libraryToLoad);
+
+        if (!dllOnDisk.exists()) {
+            extractToLocalFolder(dllOnDisk, libraryToLoad);
+        }
+
+        System.load(dllOnDisk.toString());
+
         InternalLogger.INSTANCE.trace("Successfully loaded library '%s'", libraryToLoad);
+
         return true;
     }
+
+    private static void extractToLocalFolder(File dllOnDisk, String libraryToLoad) throws IOException {
+        InputStream in = JniPCConnector.class.getClassLoader().getResourceAsStream(libraryToLoad);
+        if (in == null) {
+            throw new RuntimeException(String.format("Failed to find '%s' in jar", libraryToLoad));
+        }
+
+        OutputStream out = null;
+        try {
+            out = FileUtils.openOutputStream(dllOnDisk);
+            IOUtils.copy(in, out);
+
+            InternalLogger.INSTANCE.trace("Successfully extracted '%s' to local folder", libraryToLoad);
+        } finally {
+            if (in != null) {
+                try {
+                    in.close();
+                } catch (IOException e) {
+                    InternalLogger.INSTANCE.error("Failed to close input stream for dll extraction: %s", e.getMessage());
+                }
+            }
+            if (out != null) {
+                try {
+                    out.close();
+                } catch (IOException e) {
+                    InternalLogger.INSTANCE.error("Failed to close output stream for dll extraction: %s", e.getMessage());
+                }
+            }
+        }
+    }
+
+    private static File buildDllLocalPath() {
+        Properties properties = PropertyHelper.getSdkVersionProperties();
+        if (properties == null) {
+            throw new RuntimeException("Failed to find SDK Version Properties file.");
+        }
+
+        String version = properties.getProperty("version");
+        if (version == null) {
+            throw new RuntimeException("Failed to find SDK version.");
+        }
+
+        String dllPathName = System.getProperty("java.io.tmpdir") + File.separator + AI_BASE_FOLDER + File.separator + AI_NATIVE_FOLDER + File.separator + version;
+        File dllPath = new File(dllPathName);
+
+        if (!dllPath.exists()) {
+            dllPath.mkdirs();
+        }
+
+        if (!dllPath.exists() || !dllPath.canRead() || !dllPath.canWrite()) {
```

Comment:
Consider moving the canRead() and carWrite() checks be done on the temp folder - if you can write something there then you can probably write something under any path you create there. Also, if you can't write to the temp folder then mkdirs() is going to fail.

You should, however, keep the check that dllPath exists here.